### PR TITLE
SAMZA-1788: Add LocationIdProvider abstraction.

### PR DESCRIPTION
**Pull request by shanthoosh**
Currently in standalone, by default hostName of the standalone processor is used as LocationId. However, for containerized environments like azure cloud, kubernetes this defaulting does not work. Standalone processors can be launched from different kubernetes container on a physical machine(where each kubernetes container has different locatliyID than other kubernetes container within same machine). 

To solve this problem, we introduce locationID abstraction to allow users to plugin a uniqueId identifying the execution environment of the processor.

In containerized environments, LocationId is a composite key of multiple fields: (sliceId, containerId, hostname) By default hostname will be used as LocationId(if not configured by the user).

All the processors of an application registered from an locationID should be able to share(read/write) their local state stores. Any custom LocationIdProvider is expected to honor this contract when generating the locationID.

This patch is part of SEP-11. Please refer to it for more details.

### CODE REVIEW

**Comment by shanthoosh** [comment]
@xinyuiscool @sborya 
Please take a look when you get a chance.

File: samza-core/src/main/java/org/apache/samza/runtime/DefaultLocationIdProviderFactory.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.runtime;
+
+import org.apache.samza.util.Util;
+
+public class DefaultLocationIdProviderFactory implements LocationIdProviderFactory {
```
Comment (vjagadish1989) [inline samza-core/src/main/java/org/apache/samza/runtime/DefaultLocationIdProviderFactory.java]: Consider adding java-docs here?

File: samza-api/src/main/java/org/apache/samza/runtime/LocationIdProvider.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+import org.apache.samza.config.Config;
+
+/**
+ * Generates {@link LocationId} that uniquely identifies the
+ * execution environment of a stream processor.
+ */
+public interface LocationIdProvider {
+
+  LocationId getLocationId(Config config);
```
Comment (vjagadish1989) [inline samza-api/src/main/java/org/apache/samza/runtime/LocationIdProvider.java]: Prefer re-thinking these interfaces as follows:
```
public interface LocationIdProvider {
   LocationId getLocationId();
}
```
 
and the corresponding factory interface as:

```
public interface LocationIdProviderFactory {
  LocationIdProvider getLocationIdProvider(Config config);
}
```

Comment (shanthoosh) [reply samza-api/src/main/java/org/apache/samza/runtime/LocationIdProvider.java]: Done.

File: samza-core/src/main/java/org/apache/samza/runtime/DefaultLocationIdProviderFactory.java (hunk 1 above)
Comment (shanthoosh) [reply samza-core/src/main/java/org/apache/samza/runtime/DefaultLocationIdProviderFactory.java]: Done.

**Comment by vjagadish1989** [comment]
Looks like there are test failures. can you take a look? Otherwise, LGTM

**Pull request closed without merging**